### PR TITLE
operational-technology-use-case

### PR DESCRIPTION
**Pull request by gbatmaz**
Operational technology attack/fault detection notebook, scripts and an example model.
closes issue #36

### CODE REVIEW

File: operational-technology/training-tuning/ot-xgboost-train.py [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+
+    # Start an XGBoost classifier
+    xgb_clf = XGBClassifier(n_estimators=1000)
+
+    # Training and saving the model
+
+    xgb_clf = xgb_clf.fit(X_train, y_train)
+
+    # Save the model to a file
+    with open(modelname, "wb") as file:
```
Comment (tzemicheal) [inline operational-technology/training-tuning/ot-xgboost-train.py:97]: I wonder if you have tested the  XGBoostClassifier `save_model & load_model` method to save & restore model, instead of pickle. It could be easier to handle the versioning issue due to pickle.

File: operational-technology/inference/ot-xgboost-inference.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+    X = cudf.from_pandas(X)
+    y = cudf.from_pandas(y)
+
+    # Create train and test set
+    X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=0)
+
+    # Start an XGBoost classifier
+
+    xgb_clf = XGBClassifier()
+    with open(model, "rb") as file:
```
Comment (tzemicheal) [inline operational-technology/inference/ot-xgboost-inference.py:94]: Same as training script with load_model method instead of pickle load.

File: operational-technology/training-tuning/ot-xgboost-20230207.ipynb [hunk 3]
```diff
@@ -0,0 +1,983 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 17,
+   "id": "7e1187cb-63ce-4e39-bc50-29e881d495ca",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "xgb_clf = XGBClassifier(n_estimators=1000)"
```
Comment (tzemicheal) [inline operational-technology/training-tuning/ot-xgboost-20230207.ipynb:831]: It could be helpful if you include the training parameters of XGBoost as variables. I read you have already mentioned in the readme file (copying from it) `{'subsample': 0.8, 'n_estimators': 1200, 'min_child_weight': 2, 'max_depth': 20, 'learning_rate': 0.15, 'gamma': 0.5, 'colsample_bytree': 0.1}`

**Review by tzemicheal — CHANGES_REQUESTED** [review]
I added few comments to OT PR, thanks.

Could you add title & info of the usecase at the main [README](https://github.com/gbatmaz/morpheus-experimental/blob/ot/README.md) file for the operational technology usecase

File: operational-technology/inference/ot-xgboost-inference.py (hunk 2 above)
Comment (gbatmaz) [reply operational-technology/inference/ot-xgboost-inference.py:94]: Thank you for the feedback. When I tested, the file saved with the save_model was missing some attributes. Hence chose to use the pickle version.

File: operational-technology/training-tuning/ot-xgboost-20230207.ipynb (hunk 3 above)
Comment (gbatmaz) [reply operational-technology/training-tuning/ot-xgboost-20230207.ipynb:831]: The hyperparameter optimisation steps would take up massive space in the notebook, so we chose to leave the details in the Readme and use most of the default parameters in the code.

**Comment by gbatmaz** [comment]
> I added few comments to OT PR, thanks.
> 
> Could you add title & info of the usecase at the main [README](https://github.com/gbatmaz/morpheus-experimental/blob/ot/README.md) file for the operational technology usecase

Thank you! Just pushed a commit for this

**Review by tzemicheal — APPROVED** [review]

**Comment by tzemicheal** [comment]
LGTM!

**Review by raykallen — CHANGES_REQUESTED** [review]

File: operational-technology/README.md [hunk 4]
```diff
@@ -0,0 +1,86 @@
+## Industrial Control System (ICS) Cyber Attack Detection
+
+## Use Case
+Classify events into various categories based on power system data.
+
+### Version
+1.0
+
+### Model Overview
+The model is an XGBoost classifier that predicts each event on a power system based on dataset features.
```
[the remark this answers is not in the document]
Comment (gbatmaz) [reply operational-technology/README.md]: I have added multi-class now

File: operational-technology/README.md [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+
+Inference script can be run as:
+```
+python ot-xgboost-inference.py \
+    --model ../models/ot-xgboost-20230207.pkl \
+    --output ot-validation-output.jsonlines
+```
+This will download the dataset, the prediction is performed on the test set, and the output is saved into a file.
+
+
```
[the remark this answers is not in the document]
Comment (gbatmaz) [reply operational-technology/README.md]: Added  "how to use this model", "input data", and "output" sections to readme

File: operational-technology/README.md [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+```
+
+The hyperparameter set below came up as the best combination; different experiments may give different results.
+```
+{'subsample': 0.8, 'n_estimators': 1200, 'min_child_weight': 2, 'max_depth': 20, 'learning_rate': 0.15, 'gamma': 0.5, 'colsample_bytree': 0.1}
+```
+
+#### Model accuracy
+
+The label distribution in the dataset is not imbalanced, so we do not use the accuracy score. Instead, we use F1 weighted as the metric. The F1 score was over 0.91 on a test set.
```
[the remark this answers is not in the document]
Comment (gbatmaz) [reply operational-technology/README.md]: fixed that

**Review by raykallen — APPROVED** [review]
lgtm
